src/coreComponents/physicsSolvers/multiphysics/SinglePhasePoromechanicsKernel.hpp

```diff
@@ -378,55 +376,147 @@ class SinglePhase :
+   * @param[in] dNdX basis function derivatives
+   * @param[in] detJxW determinant of the Jacobian transformation matrix times the quadrature weight
+   * @param[inout] stack the stack variables
+   * @detail This function assembles the discretized form of the temporal derivative in the following equation
+   *   dFluidMass_dTime + divergence( fluidMassFlux ) = source  (fluid phase mass balance)
+   * with the following dependencies on the strainIncrement tensor and pressure
+   *   fluidMass = fluidMass( strainIncrement, pressure)
+   *   fluidMassFlux = fluidMassFlux( pressure)
+   */
+  GEOSX_HOST_DEVICE
+  GEOSX_FORCE_INLINE
+  void
+  assembleElementBasedFlowTerms( real64 const ( &dNdX )[numNodesPerElem][3],
+                                 real64 const & detJxW,
+                                 StackVariables & stack ) const
+  {
+    using namespace PDEUtilities;
+
+    constexpr FunctionSpace displacementTrialSpace = FE_TYPE::template getFunctionSpace< numDofPerTrialSupportPoint >();
+    constexpr FunctionSpace pressureTrialSpace = FunctionSpace::P0;
+    constexpr FunctionSpace pressureTestSpace = pressureTrialSpace;
 
-    // Compute local mass balance residual
+    // Step 1: compute local mass balance residual
     LinearFormUtilities::compute< pressureTestSpace,
                                   DifferentialOperator::Identity >
     (
       stack.localResidualMass,
       1.0,
-      fluidMassContentIncrement,
+      stack.fluidMassIncrement,
       detJxW );
 
-    // Compute local mass balance residual derivatives with respect to displacement
+    // Step 2: compute local mass balance residual derivatives with respect to displacement
     BilinearFormUtilities::compute< pressureTestSpace,
                                     displacementTrialSpace,
                                     DifferentialOperator::Identity,
                                     DifferentialOperator::Divergence >
     (
       stack.dLocalResidualMass_dDisplacement,
       1.0,
-      dFluidMassContent_dVolStrainIncrement,
+      stack.dFluidMassIncrement_dVolStrainIncrement,
       dNdX,
       detJxW );
 
-    // Compute local mass balance residual derivatives with respect to pressure
+    // Step 3: compute local mass balance residual derivatives with respect to pressure
     BilinearFormUtilities::compute< pressureTestSpace,
                                     pressureTrialSpace,
                                     DifferentialOperator::Identity,
                                     DifferentialOperator::Identity >
     (
       stack.dLocalResidualMass_dPressure,
       1.0,
-      dFluidMassContent_dPressure,
+      stack.dFluidMassIncrement_dPressure,
       1.0,
       detJxW );
   }
 
-  /**
-   * @copydoc geosx::finiteElement::ImplicitKernelBase::complete
-   */
+  GEOSX_HOST_DEVICE
+  GEOSX_FORCE_INLINE
+  void quadraturePointKernel( localIndex const k,
```

Comment:
I think this functions is the same in `src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp` so maybe it can be moved to the base class ?

Reply:
I guess it's the duplication question we spoke about last time?

Reply:
Yes, I think we can get rid of this duplication as well, see [this comment](https://github.com/GEOSX/GEOSX/pull/2194#discussion_r1040016288). I will give it a try and we can see if we like it.